cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -1052,6 +1069,47 @@ protected Future<Void> updateKafkaVersion(KafkaStatus kafkaStatus) {
         return Future.succeededFuture();
     }
 
+    protected Future<Void> updateKafkaMetadataState(KafkaStatus kafkaStatus) {
+        kafkaStatus.setKafkaMetadataState(kafkaMetadataStateManager.computeNextMetadataState(kafkaStatus).name());
+        return Future.succeededFuture();
+    }
+
+    protected Future<Void> maybeKafkaMetadataMigrationInProgress() {
```

Comment:
I somehow don't understand what this is doing. It is called at the end of the reconciliation, and it does nto seem to do anything really, so what is its purpose? Maybe some Javadoc would be good to explain it?

Reply:
Why do you think it's not doing anything? I will add Javadoc but this method, if migration is ongoing (based on the status), checks if migration is still going on or it finished on the Kafka side. It's calling the `KafkaMetadataStateManager.checkMigrationInProgress` which leverages the Kafka Agent to know if the migration ended (via corresponding Kafka metric).

Reply:
The name `checkMigrationInProgress` suggests it doesn't _change_ anything, it just checks some state. But that seems to be useless at the end of the reconciliation where it does not seem to be used for anything anyway.

Reply:
The `checkMigrationInProgress` is actually changing an internal flag in the `KafkaMetadataStateManager` if the migration is done. That flag, `isMigrationDone` is then used for the evolution of the FSM to move to the next "dual-write" state.

Reply:
That seems weird to me:
* You probably want to check it at the start of the next reconciliation and not at the end of the previous one? That way you have the latest up-to-date information.
* You do not seem to update any state. You do not even use any output from the `checkMigrationInProgress` method. So where and how do you know what the state should be?

Reply:
So the `isMigrationDone` is set here https://github.com/strimzi/strimzi-kafka-operator/blob/zk-kraft-migration/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java#L170 when the Kafka Agent metric reports the Kafka migration is done.

Then it's used here https://github.com/strimzi/strimzi-kafka-operator/blob/zk-kraft-migration/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java#L225
which it allows to move the FSM from migration to dual-write (called by the `computeNextMetadataState` in the same class).

The `computeNextMetadataState` is called at the end of the reconcile to update the migration state on the Kafka CR https://github.com/strimzi/strimzi-kafka-operator/blob/zk-kraft-migration/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java#L1071

Reply:
> So the isMigrationDone is set here https://github.com/strimzi/strimzi-kafka-operator/blob/zk-kraft-migration/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java#L170 when the Kafka Agent metric reports the Kafka migration is done.

I guess you should then not call the method `checkMigrationInProgress`. `check` suggests it checks something and possibly returns some result and not that it updates anything. You should probably call it `maybeUpdateMigrationState` or something. That will make it clear that it changes something and not just checks something.

> The computeNextMetadataState is called at the end of the reconcile to update the migration state on the Kafka CR https://github.com/strimzi/strimzi-kafka-operator/blob/zk-kraft-migration/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java#L1071

Why is it called at the end of the reconciliation?
* It does not seem to be really used for anything anymore in this reconciliation but only in the next one.
* The actual state at the start of the next reconciliation might have changed in the _up to 2 minutes_. So isn't it worth it to get and use a fresh state?

Reply:
>  It does not seem to be really used for anything anymore in this reconciliation but only in the next one.

It's used for triggering a new reconcile, if the `status` in the `Kafka` CR is changed and the migration FSM needs to take some action from there. Also, how the FSM works is ... 1. get the current status at the beginning of the reconciliation (when the `KafkaReconciler` object is created), 2. take action, 3. update with the new status in the `Kafka` CR. So covering the state transition.

> The actual state at the start of the next reconciliation might have changed in the up to 2 minutes. So isn't it worth it to get and use a fresh state?

The actual state can change only because of the `computeNextMetadataState` call, so I don't get what you mean.

Reply:
> The actual state can change only because of the computeNextMetadataState call, so I don't get what you mean.

Isn't the actual state in the broker? As that seems to be what you are checking through the agent.

Reply:
What we are checking via the agent is "only" the migration process itself, so when Kafka controllers are getting data from ZooKeeper copying them to be KRaft. What we have in the `status` of the `Kafka` CR is the state of the overall FSM driving the migration (of course during actual data copying, "migration" on Kafka will mean `KRaftMigration` in the FSM).